src/Syntax.hs

```diff
@@ -1,145 +1,135 @@
-type Pn = AlexPosn
 
 type Id = String
 
-data Act = Main [RawBehaviour]
+-- can't override show instance for AlexPosn
+-- so seems reasonable to use our own
+data Pn = Pn Int Int
+
+data Act = Act RawConstructor [RawBehaviour]
   deriving (Eq, Show)
 
-data RawBehaviour
-    = Transition Id Id Interface [IffH] (Case Expr Post) (Maybe Ensures)
-    | Constructor Id Id Interface [IffH] Creates [ExtStorage] (Maybe Ensures) (Maybe Invariants)
+data RawConstructor = RawConstructor
+  Id -- name
+  Id -- contract
+  Interface
+  [Creation] -- creates
+  -- missing return
   deriving (Eq, Show)
 
-type Ensures = [Expr]
+data RawBehaviour = RawBehaviour
+  Id -- name
+  Id -- contract
+  Interface
+  (Maybe [Expr]) -- preconditions
+  [Case] -- cases
+  deriving (Eq, Show)
 
-type Invariants = [Expr]
+-- `uint a := 1`
+-- `mapping(address => uint)`
+data Creation
+  = CDefn Defn
+  | CDecl Decl
+  deriving (Eq, Show)
 
 data Interface = Interface Id [Decl]
   deriving (Eq)
 
-instance Show Interface where
-  show (Interface a d) = a <> "(" <> intercalate ", " (fmap show d) <> ")"
-
-data Case a b
-    = Leaf Pn a b
-    | Branch Pn a [Case a b]
+data Case = Case Expr [Claim]
   deriving (Eq, Show)
 
-data Post
-    = Post (Maybe [Storage]) [ExtStorage] (Maybe Expr)
+data Claim
+  = StorageClaim [Store]
+  | ReturnClaim Expr
   deriving (Eq, Show)
 
-data Creates
-    = Creates [Assign]
+data Store = Store Ref Expr
   deriving (Eq, Show)
 
-data Storage = Rewrite Entry Expr
-             | Constant Entry
+data Ref
+  = Ref Id
+  | Zoom Id Expr
   deriving (Eq, Show)
 
-data ExtStorage
-    = ExtStorage Id [Storage]
-    | ExtCreates Id Expr [Assign]
-    | WildStorage
-  deriving (Eq, Show)
+data Decl = Decl Type Id
+  deriving (Eq)
 
-data Assign = AssignVal StorageVar Expr | AssignMany StorageVar [Defn] | AssignStruct StorageVar [Defn]
+data Defn = Defn Type Id Expr
   deriving (Eq, Show)
 
-data IffH = Iff Pn [Expr] | IffIn Pn AbiType [Expr]
+data Type
+  = AbiType AbiType
+  | Mapping Type Type
   deriving (Eq, Show)
 
-data Entry
-  = Entry Pn Id [Expr]
-  | Wild
-  deriving (Eq, Show)
+data Expr
 
---data Defn = Defn Pn Expr Expr
-data Defn = Defn Expr Expr
-  deriving (Eq, Show)
+  -- booleans
+  = EBoolLit Bool       -- `true`
+  | EAnd Expr Expr      -- `a and b`
+  | EOr Expr Expr       -- `a or b`
+  | ENot Expr           -- `not a`
+  | EEq Expr Expr       -- `a == b`
+  | ENeq Expr Expr      -- `a =/= b`
+  | ELE Expr Expr       -- `a <= b`
+  | ELT Expr Expr       -- `a < b`
+  | EGE Expr Expr       -- `a >= b`
+  | EGT Expr Expr       -- `a > b`
+
+  -- numbers
+  | EIntLit Integer     -- `666`
+  | EAdd Expr Expr      -- `a + b`
+  | ESub Expr Expr      -- `a - b`
+  | EMul Expr Expr      -- `a * b`
+  | EDiv Expr Expr      -- `a / b`
+  | EMod Expr Expr      -- `a % b`
+  | EExp Expr Expr      -- `a ^ b`
+
+  -- other
+  | ERead Ref           -- `a`
+  | EEnv EthEnv         -- `CALLVALUE`
+  | EITE Expr Expr Expr -- `if a then b else c`
```

Comment:
with `fi` or no? You added its lexeme. If `fi` can be avoided I would like to avoid it

Reply:
Yeah, I wanted to ask you about this. It can be avoided. Without the `fi`, the ambiguous `if a then b else c + d` will always parse as `if a then b else (c + d)` rather than `(if a then b else c) + d`, which seems like a reasonable norm.

Reply:
I agree